File: app/src/main/java/ru/tubetv/app/SearchClient.java
package ru.tubetv.app;

import android.net.Uri;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedInputStream;
import java.io.ByteArrayOutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

final class SearchClient {
    private static final int LIMIT = 12;
    private static final int TIMEOUT_MS = 7000;

    List<VideoItem> searchRutube(String query, int minWidth) throws Exception {
        String url = "https://rutube.ru/api/search/video/?query="
                + Uri.encode(query) + "&page=1";
        JSONObject root = getJson(url);
        JSONArray items = root.optJSONArray("results");
        List<VideoItem> result = new ArrayList<>();
        if (items == null) return result;
        for (int i = 0; i < Math.min(items.length(), LIMIT); i++) {
            JSONObject item = items.getJSONObject(i);
            String id = item.optString("id");
            String page = item.optString("video_url", "https://rutube.ru/video/" + id + "/");
            String embed = item.optString("embed_url", "https://rutube.ru/play/embed/" + id);
            embed = copyQueryParameter(page, embed, "p");
            int duration = item.optInt("duration");
            result.add(new VideoItem("RUTUBE", item.optString("title", "Без названия"),
                    "", item.optString("thumbnail_url"),
                    embed, page, duration * 1000L));
        }
        if (minWidth <= 0 || result.isEmpty()) return result;
        return filterByQuality(result, minWidth);
    }

    List<VideoItem> searchVk(String query, int minWidth) throws Exception {
        return searchVk(query, minWidth, 480);
    }

    List<VideoItem> searchVk(String query, int minWidth, int thumbnailWidth) throws Exception {
        return new VkWebClient().search(query, minWidth, thumbnailWidth);
    }

    int searchVkPages(String query, int minWidth, int thumbnailWidth,
                      VkWebClient.PageListener listener) throws Exception {
        return new VkWebClient().searchPages(query, minWidth, thumbnailWidth, listener);
    }

    List<VideoItem> searchDzen(String query, int minWidth) throws Exception {
        return new DzenClient().search(query, minWidth);
    }

    List<VideoItem> searchOk(String query, int minWidth) throws Exception {
        return OkClient.search(query, minWidth);
    }

    List<VideoItem> searchPeerTube(String query, int minWidth, int thumbnailWidth) throws Exception {
        return PeerTubeClient.search(query, minWidth, thumbnailWidth);
    }

    int searchDzenPages(String query, DzenClient.PageListener listener) throws Exception {
        return new DzenClient().searchPages(query, listener);
    }

    static List<VideoItem> filterByQuality(List<VideoItem> candidates, int minWidth) throws Exception {
        return filterByQuality(candidates, minWidth, 4);
    }

    static List<VideoItem> filterByQuality(List<VideoItem> candidates, int minWidth,
                                           int maxParallelRequests) throws Exception {
        List<VideoItem> inspected = inspectQualities(candidates, maxParallelRequests);
        List<VideoItem> filtered = new ArrayList<>();
        for (VideoItem item : inspected) if (item.maxWidth >= minWidth) filtered.add(item);
        return filtered;
    }

    private static String copyQueryParameter(String source, String target, String name) {
        if (source == null || target == null) return target;
        try {
            String value = Uri.parse(source).getQueryParameter(name);
            if (value == null || value.isEmpty()
                    || Uri.parse(target).getQueryParameter(name) != null) return target;
            return Uri.parse(target).buildUpon().appendQueryParameter(name, value).build().toString();
        } catch (Exception ignored) {
            return target;
        }
    }

    static List<VideoItem> inspectQualities(List<VideoItem> candidates, int maxParallelRequests) {
        if (candidates.isEmpty()) return new ArrayList<>();
        int workers = Math.max(1, Math.min(maxParallelRequests, candidates.size()));
        ExecutorService probes = Executors.newFixedThreadPool(workers);
        try {
            List<Future<VideoItem>> futures = new ArrayList<>();
            for (VideoItem item : candidates) {
                futures.add(probes.submit(() -> {
                    PlaybackInfo info = new StreamResolver().inspect(item.playUrl);
                    return item.withQuality(info.maxWidth, info.maxHeight);
                }));
            }
            List<VideoItem> inspected = new ArrayList<>();
            for (Future<VideoItem> future : futures) {
                if (Thread.currentThread().isInterrupted()) break;
                try {
                    VideoItem item = future.get();
                    if (item != null) inspected.add(item);
                } catch (Exception ignored) { }
            }
            return inspected;
        } finally {
            probes.shutdownNow();
        }
    }

    private static JSONObject getJson(String address) throws Exception {
        HttpURLConnection connection = (HttpURLConnection) new URL(address).openConnection();
        connection.setConnectTimeout(TIMEOUT_MS);
        connection.setReadTimeout(TIMEOUT_MS);
        connection.setRequestProperty("Accept", "application/json");
        connection.setRequestProperty("User-Agent", "0W-Tube/0.5.14 AndroidTV");
        try {
            int code = connection.getResponseCode();
            if (code < 200 || code >= 300) throw new Exception("HTTP " + code);
            BufferedInputStream input = new BufferedInputStream(connection.getInputStream());
            ByteArrayOutputStream output = new ByteArrayOutputStream(32 * 1024);
            byte[] buffer = new byte[8192];
            int read;
            while ((read = input.read(buffer)) != -1) output.write(buffer, 0, read);
            return new JSONObject(output.toString(StandardCharsets.UTF_8.name()));
        } finally {
            connection.disconnect();
        }
    }

}

File: MacOS/Sources/App/Clients/SearchClient.swift
import Foundation

/// Фасад над бэкендами и общий механизм догрузки качества.
enum SearchClient {
    static func searchRutube(query: String, minWidth: Int) async throws -> [VideoItem] {
        try await RutubeClient.search(query: query, minWidth: minWidth)
    }

    static func searchVk(query: String, minWidth: Int, thumbnailWidth: Int = 480) async throws -> [VideoItem] {
        try await VkWebClient.search(query: query, minWidth: minWidth, thumbnailWidth: thumbnailWidth)
    }

    static func searchVkPages(query: String, minWidth: Int, thumbnailWidth: Int,
                              onPage: ([VideoItem]) async -> Bool) async throws -> Int {
        try await VkWebClient.searchPages(query: query, minWidth: minWidth,
                                          thumbnailWidth: thumbnailWidth, onPage: onPage)
    }

    static func searchDzen(query: String, minWidth: Int) async throws -> [VideoItem] {
        try await DzenClient.search(query: query, minWidth: minWidth)
    }

    static func searchDzenPages(query: String, onPage: ([VideoItem]) async -> Bool) async throws -> Int {
        try await DzenClient.searchPages(query: query, onPage: onPage)
    }

    static func searchOk(query: String, minWidth: Int) async throws -> [VideoItem] {
        try await OkClient.search(query: query, minWidth: minWidth)
    }

    static func searchPeerTube(query: String, minWidth: Int, thumbnailWidth: Int) async throws -> [VideoItem] {
        try await PeerTubeClient.search(query: query, minWidth: minWidth, thumbnailWidth: thumbnailWidth)
    }

    static func filterByQuality(_ candidates: [VideoItem], minWidth: Int,
                                maxParallelRequests: Int = 4) async throws -> [VideoItem] {
        let inspected = await inspectQualities(candidates, maxParallelRequests: maxParallelRequests)
        return inspected.filter { $0.maxWidth >= minWidth }
    }

    /// Догружает реальное максимальное разрешение для карточек, которые источник отдал
    /// без него. Число одновременных проверок ограничено: сервисы плохо реагируют на
    /// пачку запросов к плееру, а карточки уже показаны и обновляются на месте.
    /// Ролики, чей поток проверить не удалось, из результата выпадают и сохраняют
    /// прежнее (неизвестное) качество.
    static func inspectQualities(_ candidates: [VideoItem], maxParallelRequests: Int) async -> [VideoItem] {
        if candidates.isEmpty { return [] }
        let workers = max(1, min(maxParallelRequests, candidates.count))
        return await withTaskGroup(of: VideoItem?.self) { group in
            var index = 0
            var result: [VideoItem] = []
            func addNext() {
                guard index < candidates.count else { return }
                let item = candidates[index]
                index += 1
                group.addTask {
                    guard let info = try? await StreamResolver.inspect(item.playUrl) else { return nil }
                    return item.withQuality(info.maxWidth, info.maxHeight)
                }
            }
            for _ in 0..<workers { addNext() }
            while let finished = await group.next() {
                if Task.isCancelled { group.cancelAll() }
                if let finished { result.append(finished) }
                addNext()
            }
            return result
        }
    }
}
